File: testdata/fixtures/swift/CWE-798.swift
// CWE-798: the credential ships inside the .ipa.
//
// An App Store binary is a zip file, and `unzip` plus `strings` is the whole
// of the work - the string pool of the compiled binary holds every literal the
// app was built with. A key written into a `let` is a key that everyone who
// installed the app already has, so rotating it means shipping a release and
// waiting for people to update.
//
// What separates a credential from a name is the value rather than the name.
// `"password"` beside `"host"` is a keyword table; `"API_KEY"` is what the
// value is called somewhere else; `"$(AUTH_TOKEN)"` is a build setting, which
// is the fix.
import Foundation

enum Config {

    // ruleid: swift-credential-is-a-literal
    static let apiKey = "AIzaSyD3vK8pQ2mN7xR1tY9wZ4bC6eF0gH5jL2k"

    // ruleid: swift-credential-is-a-literal
    private let dbPassword = "hunter2"

    // ruleid: swift-credential-is-a-literal
    static let adminSecret: String = "s3cr3t-admin-2019"

    // ruleid: swift-credential-is-a-literal
    var authToken = "ghp_16C7e42F292c6912E7710c838347Ae178B4a"

    // ok: swift-credential-is-a-literal
    static let apiKeyDefaultsName = "com.example.apiKey"

    // ok: swift-credential-is-a-literal
    static let passwordField = "password"

    // ok: swift-credential-is-a-literal
    static let tokenPlaceholder = "$(AUTH_TOKEN)"

    // ok: swift-credential-is-a-literal
    static let apiKey2 = ProcessInfo.processInfo.environment["API_KEY"]

    // ok: swift-credential-is-a-literal
    static let emptyPassword = ""

    // ok: swift-credential-is-a-literal
    static let authBaseURL = "https://api.example.com/v2"

    // ok: swift-credential-is-a-literal
    static let tokenPattern = "^[A-Za-z0-9_-]{16,}$"
}

File: testdata/fixtures/kotlin/CWE-798.kt
// CWE-798: the credential is in the APK.
//
// An Android app is a zip file on a device the attacker owns. `apktool d` and
// a grep is the whole of the work, and a key compiled into a `const val` is a
// key in every copy of the app that was ever installed - so rotating it means
// shipping a release and waiting for people to take it.
//
// What separates a credential from a name is the value rather than the name.
// `"password"` beside `"host"` is a keyword table; `"API_KEY"` is what the
// value is called somewhere else; `"\${api.key}"` is the fix, not the bug.
package fixture

object Config {

    // ruleid: kotlin-credential-is-a-literal
    private const val API_KEY = "AIzaSyD3vK8pQ2mN7xR1tY9wZ4bC6eF0gH5jL2k"

    // ruleid: kotlin-credential-is-a-literal
    private val dbPassword = "hunter2"

    // ruleid: kotlin-credential-is-a-literal
    const val ADMIN_SECRET: String = "s3cr3t-admin-2019"

    // ok: kotlin-credential-is-a-literal
    private const val API_KEY_PROPERTY = "com.example.apiKey"

    // ok: kotlin-credential-is-a-literal
    private const val PASSWORD_FIELD = "password"

    // ok: kotlin-credential-is-a-literal
    private const val TOKEN_PLACEHOLDER = "\${auth.token}"

    // ok: kotlin-credential-is-a-literal
    private val apiKey = BuildConfig.API_KEY

    // ok: kotlin-credential-is-a-literal
    private val emptyPassword = ""

    // ok: kotlin-credential-is-a-literal
    const val BASE_URL = "https://api.example.com"
}
